refactor(devtools): styles management

hawkgs opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?

SCSS files are scattered across a few directories. There is some duplication as well.

## What is the new behavior?

1. Move all styles in a single folder as close as possible to `ng-devtools` component – as stated in the section above there had been SCSS files that resided in the app wrappers (the shell and demo apps) or the main `devtools/` directory. There were also duplicated browser-specific SCSS files in each app wrapper. This change should, hopefully, make style management a bit easier. Each app wrapper still retains a `styles.scss` which should ultimately be used only for applying styles that are specific to that app wrapper, not the devtools.

2. Set a browser class to the document body – this is an improvement of #46037. Basically, additionally to loading browser-specific global styles, we also set a browser-specific class to the body. We can use that for any granular or component-based styles.

3. SCSS mixins – right now the theme styles are applied via `:host-context(<THEME-CLASS>)`. What the mixins do is to encapsulate that logic in case it changes (or the class names change). The mixins also provide a parameter for targeting a specific browser. Eventually, some of these mixins might render useless, so we can drop them.

**Next changes**

- Converting to the theme mixins
- Extracting some of the colors and creating actual themes based on them

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

hawkgs committed:
refactor(devtools): styles management

- Move all styles to ng-devtools/src/styles.
- Create a BrowserService that detects the browsers and adds it as a class to the body. Move global browser styles.
- Create theme mixins that incorporate the browser type into them.
- Refactor some of the affected code along with the introduced changes.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/application-services/browser-styles-service.ts:
```diff
@@ -0,0 +1,78 @@
+  private _browser: Browser = 'unknown';
+
+  initBrowserSpecificStyles() {
+    this._browser = this._detectBrowser();
```
Suggestion: Move this to the field initializer so we don't have to worry about remembering to call `initBrowserSpecificStyles` before any other functionality in this class.

```typescript
class BrowserStylesService {
  private browser = this.detectBrowser();

  // ...
}
```

hawkgs replied:
Makes sense.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/application-services/browser-styles-service.ts:
```diff
@@ -0,0 +1,78 @@
+    }
+    return 'unknown';
+  }
+}
```
Consider: It would be helpful to include some basic unit tests for `BrowserStylesService` and `ThemeService`. I get this is mostly refactoring existing code, so I don't think tests are a hard blocker here, but these classes seem relatively simple and I'd like to improve our test coverage wherever we can.

dgp1130 reviewed (commented):
I left some general feedback, but I'm mostly happy to defer to @AleksanderBodurri on the specifics.

hawkgs commented on devtools/projects/ng-devtools/src/lib/application-services/browser-styles-service.ts:
```diff
@@ -0,0 +1,78 @@
+  private readonly _doc = inject(DOCUMENT);
+  private readonly _rendererFactory = inject(RendererFactory2);
+  private readonly _renderer = this._rendererFactory.createRenderer(null, null);
+  private readonly _platform = inject(Platform);
```
Okay. I am a fan of that naming convention and since I noticed that some of the files use it I went with it. I'll rename the props. :+1:

hawkgs commented on devtools/projects/ng-devtools/src/lib/application-services/browser-styles-service.ts:
```diff
@@ -0,0 +1,78 @@
+};
+
+const BROWSER_STYLES: {[key in Browser]: string} = {
+  'chrome': 'chrome.css',
```
Although they are similar, Edge for example, uses a bit different UI elements and colors.

hawkgs committed:
fixup! refactor(devtools): styles management

- Address reviewer's comments
- Introduce `application-providers`
- Test `BrowserStylesService` and `ThemeService`

hawkgs commented:
Done. I did some additional changes like the introduction of a `window` provider since I noticed that we are using the global object directly. I needed it for mocking some of the code for the `ThemeService`. Also, went from kebab to snake case for service names.

AleksanderBodurri commented on devtools/projects/ng-devtools/src/lib/devtools.component.ts:
```diff
@@ -53,6 +54,7 @@ const LAST_SUPPORTED_VERSION = 9;
     ]),
   ],
   imports: [DevToolsTabsComponent, MatTooltip, MatProgressSpinnerModule, MatTooltipModule],
+  providers: [WINDOW_PROVIDER, ThemeService],
```
Is WINDOW_PROVIDER required in this file? I don't see an inject for Window

AleksanderBodurri replied:
Ahh nvm I see the inject in theme service

AleksanderBodurri commented on devtools/projects/ng-devtools/src/lib/application-services/browser_styles_service.ts:
```diff
@@ -0,0 +1,76 @@
+
+// Keep class names in sync with _theme.scss and _browser.scss
+const BROWSER_CLASS_NAME: {[key in Browser]: string} = {
+  'chrome': 'chrome-ui',
```
+1 on @dgp1130  suggestion to add edge to this list. It's more similar to chrome than firefox but the UI is still distinct

hawkgs replied:
Okay. Generally, I don't mind adding it. It just that I've added what's supported so far.

AleksanderBodurri reviewed (commented):
Looks great overall, left a couple of comments. Thanks for this @hawkgs  😄

AleksanderBodurri reviewed (approved):
Looks like this is missing some headers in new files (see tslint failure in Ci) but otherwise LGTM

hawkgs commented:
My bad, sorry. Fixed.

atscott commented:
This PR was merged into the repository by commit 4f3ad984669c554b90e49fbe9b3795d95033780f.

The changes were merged into the following branches: main
